BOLT 12 Offers utilities

jkczyz opened the pull request:
Add utility methods to `ChannelManager` for:
- creating `OfferBuilder` such that derived keys are used for the signing pubkey,
- creating `RefundBuilder` such that derived keys are used for the payer id,
- sending an `InvoiceRequest` for an `Offer` such that derived keys are used for the payer id, and
- TODO: sending a `Bolt12Invoice` for a `Refund` such that derived keys are used for the signing pubkey

When the `ChannelManager` is the payer (i.e., when creating `Refund` and `InvoiceRequest`), a `PaymentId` is required and used for tracking the outbound payment once the `Bolt12Invoice` is received.

One-hop blinded paths are used throughout until multi-hop blinded paths are supported.

Offer message handling is in #2039, which is based on this PR.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6818,6 +6829,121 @@ where
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	/// [`InvoiceRequestBuilder`]: crate::offers::invoice_request::InvoiceRequestBuilder
+	/// [`Bolt12Invoice`]: crate::offers::invoice::Bolt12Invoice
+	pub fn request_invoice(
```
Ah, this is the "paying for an offer" method. I guess that I can rename it to be more user- rather than implementation-focused.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -155,6 +155,28 @@ where
+	/// The message contents to send in an [`OnionMessage`].
+	///
+	/// [`OnionMessage`]: crate::ln::msgs::OnionMessage
+	pub contents: T,
```
We'll need a more general bound since `OffersMessage` is used, too.

jkczyz commented on lightning/src/onion_message/packet.rs:
```diff
@@ -118,6 +118,13 @@ pub enum OnionMessageContents<T: CustomOnionMessageContents> {
 }
 
+/// A type used within a variant of [`OnionMessageContents`].
+pub trait OnionMessageContentsType {}
```
But we don't want a BOLT 12 handler to be able to send arbitrary ones. A custom handler can alway have their type wrap `OffersMessage` in a variant.

jkczyz replied:
Discussed offline and made the following changes:
- Renamed `OnionMessageContents` to `ParsedOnionMessageContents` and gave it `pub(super)` visibility
- Renamed `CustomOnionMessageContents` to `OnionMessageContents` and used it as a trait bound on most methods
- Changed Payload to hold `T` instead of `ParsedOnionMessageContents<T>` where `T: OnionMessageContents`
- Implemented `OnionMessageContents` for `OffersMessage` and `CustomOnionMessageContents`. The latter is needed so that arbitrary `Payload` contents can be read and specific contents can be written

valentinewallace reviewed (commented):
Initial look, just nits

jkczyz commented on fuzz/src/onion_message.rs:
```diff
@@ -208,19 +211,6 @@ mod tests {
 
 	#[test]
 	fn test_no_onion_message_breakage() {
-		let one_hop_om = "020000000000000000000000000000000000000000000000000000000000000e01055600020000000000000000000000000000000000000000000000000000000000000e01ae0276020000000000000000000000000000000000000000000000000000000000000002020000000000000000000000000000000000000000000000000000000000000e0101022a0000000000000000000000000000014551231950b75fc4402da1732fc9bebf00109500000000000000000000000000000 ...
```
Restored it but needs to check against `InvalidFirstHop`, which I'm not sure if that is intended?

jkczyz replied:
Will reexamine this but want to get the other changes pushed for now.

valentinewallace replied:
Sure, I'd also be fine with saving this for follow-up if you want to note it in #1970.

jkczyz replied:
Ok, other than the introduction node being wrong, when hardcoding a fix I'm also seeing this logging:

```
Received an onion message with path_id Some([]) and a reply_path
```

Instead of the expected:

```
Received an onion message with path_id None and a reply_path
```

But I don't understand how my change would affect the `path_id`. Any ideas?

jkczyz replied:
Oh, nevermind on this. That was just because I had been building using some unpublished commits instead of during the rebase. 🤦‍♂️

jkczyz replied:
Opened #2671

jkczyz commented on lightning/src/blinded_path/mod.rs:
```diff
@@ -66,7 +66,7 @@ impl BlindedPath {
 	/// pubkey in `node_pks` will be the destination node.
 	///
-	/// Errors if less than one hop is provided or if `node_pk`(s) are invalid.
+	/// Errors if no hop are provided or if `node_pk`(s) are invalid.
```
Fixed locally, but will pluralize "hops" the next time I need to push.

benthecarman commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6918,6 +7060,15 @@ where
+		let entropy_source = self.entropy_source.deref();
+		let secp_ctx = &self.secp_ctx;
+		BlindedPath::one_hop_for_message(self.get_our_node_id(), entropy_source, secp_ctx).unwrap()
+
```
nit: extra line here

jkczyz commented:
Rebased as there were a lot of merge conflicts from the last release.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7067,6 +7078,139 @@ where
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	pub fn create_offer_builder(
+		&self, description: String
+	) -> OfferBuilder<DerivedMetadata, secp256k1::All> {
```
May want to set a sane default for the expiry. It may also be nice in the future to add accessor methods on `OfferBuilder` so users can tell what we've set.

valentinewallace replied:
BOLT 11 invoices default to 1 hour expiry if none is specified, so it may be confusing that there's a different default behavior here. A comment may be useful but fine if you don't think it'll be confusing.

jkczyz replied:
Yeah, BOLT 12 invoices also have a default expiry, but offers do not. I'd prefer to at least not do anything to surprising with regard to the spec, at least. Let me update the docs.

valentinewallace commented on lightning/src/onion_message/offers.rs:
```diff
@@ -32,7 +34,17 @@ const INVOICE_ERROR_TLV_TYPE: u64 = 68;
+	///
+	/// The returned [`OffersMessage`], if any, is enqueued to be sent by [`OnionMessenger`].
+	///
+	/// [`OnionMessenger`]: crate::onion_message::messenger::OnionMessenger
```
```suggestion
	/// [`OnionMessenger`]: crate::onion_message::OnionMessenger
```
Fixes `cargo doc`

jkczyz reviewed (commented):
Need to look into addressing the fuzz test comment still.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7067,6 +7078,152 @@ where
+			.path(path);
+
+		self.pending_outbound_payments
+			.add_new_awaiting_invoice(payment_id, retry_strategy, max_total_routing_fee_msat)
```
Is it the case that if this refund is not paid within 3 timer ticks, we won't accept invoices to it even if no expiry is set?

jkczyz replied:
Hmm... yeah, we were thinking more of the invoice request scenario where the response should be immediate. I think we'll want to add an expiration parameter and use that to determine how many timer ticks are allowed for refunds.

For the payment, we'll at least check the expiry in `OutboundPayments::find_route_and_send_payment`, but only in `std`. In general, it seems we'll always attempt to pay expired invoices in `no-std`, IIUC. 😕

jkczyz replied:
> Hm, relying on timer ticks to determine expiration sounds a bit error-prone since `timer_tick_occurred` is called "roughly 1 per minute". Maybe we should pass in the current time into `timer_tick_occurred` in `no-std` or something like that?

That's a fair point, though I'm not 100% certain how concerned we should be. We are fuzzy elsewhere with time where we are receiving money. But here we are sending money, so maybe we should be? Worse case scenario is we pay someone after user said we shouldn't. But that's only if the user wasn't calling `timer_ticker_occurred` appropriately. @TheBlueMatt Any thoughts on this?

So would we store the refund's absolute expiry in `AwaitingInvoice` and then in `std` presumably rely on `SystemTime`? For an invoice request for an offer, would we just use a hardcoded 3-minute expiration before forgetting about the request?

Note that we won't pay the invoice if _it_ is expired, not withstanding the earlier caveat about `no-std`. But it just occurred to me that my earlier statement was not exactly accurate:

> For the payment, we'll at least check the expiry in `OutboundPayments::find_route_and_send_payment`, but only in `std`. In general, it seems we'll always attempt to pay expired invoices in `no-std`, IIUC. 😕

Here, we would check the expiry of the invoice -- not the refund -- which is controlled by the refund recipient. The spec doesn't give any guidance on what to do if the refund is expired but the invoice is not. Should get this cleared up.

valentinewallace replied:
I was thinking in the `Refund` case, if a merchant wants to allow a user a year to claim a refund, but they call `timer_tick_occurred` a bit too often then the user may only have like 8 months to claim in practice. But since the fix seems potentially annoying, it's probably fine to wait and see if issues actually arise?

jkczyz replied:
Yeah, the refund case is the concern. Customer requests a refund from the merchant and the merchants emails a QR code. Customer scans the QR code and sends an invoice. The time between emailing the refund QR and the customer sending the invoice could be large. It's a much different use case then when the customer scans an offer.

jkczyz replied:
Discussed offline. Updated to use `SystemTime` instead of timer ticks and block time if `no-std`. We also discussed a timer-based approach for shorter-lived refunds (i.e., less than a day) when building with `no-std`. The latest push only contains the the aforementioned approach for simplicity and to get quicker feedback. I'm not 100% clear on how we'd want to combine the timer-based approach in `AwaitingInvoice`. @TheBlueMatt Were you thinking simply having an `Option<64>` for remaining ticks? If `Some` use this in lieu of the absolute expiry?

jkczyz replied:
Hmm... ticks are essentially number of minutes left until it expires whereas the expiry is a fixed time when it expires. So they can't be compared directly unless I'm misunderstanding how they should both be represented.

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -456,7 +485,7 @@ where
 	/// Handle an incoming onion message. Currently, if a message was destined for us we will log, but
 	/// soon we'll delegate the onion message to a handler that can generate invoices or send
 	/// payments.
```
This is outdated. It can probably be removed since docs on trait implementations are fairly hidden

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -598,19 +633,27 @@ where
+	// node, and then enqueue the message for sending to the first peer in the full path.
+	fn next_onion_message_for_peer(&self, peer_node_id: PublicKey) -> Option<OnionMessage> {
+		// Enqueue any initiating `OffersMessage`s to send.
+		for message in self.offers_handler.release_pending_messages() {
+			let PendingOnionMessage { contents, destination, reply_path } = message;
+			self.find_path_and_enqueue_onion_message(
+				contents, destination, reply_path, format_args!("when sending OffersMessage")
+			);
+		}
+
+		// Enqueue any initiating `CustomMessage`s to send.
+		for message in self.custom_handler.release_pending_custom_messages() {
+			let PendingOnionMessage { contents, destination, reply_path } = message;
+			self.find_path_and_enqueue_onion_message(
+				contents, destination, reply_path, format_args!("when sending CustomMessage")
+			);
+		}
```
I chatted with @TheBlueMatt about this in person, awhile back. IIRC, this is essentially a user-initiated action, so it would be rare. The alternative would be needing to initiate path finding in handler implementations, which would require them having a router.

But you're right that it could lead to a delay in message processing. @TheBlueMatt Is ten messages too many? Should we re-consider calling path finding here?

valentinewallace replied:
Ah okay, yeah I suppose it *is* a user-initiated action so it should be rare. Just struck me as weird when reading it.

valentinewallace commented:
This is good for rebase, I'd be happy with a squash too.

jkczyz commented:
> Landed #2662 so this is good for rebase, I'd be happy with a squash too.

Rebased, squashed, and included potential fixups for the `Refund` timer ticks based on absolute expiry for consideration.

jkczyz commented:
> Rebased, squashed, and included potential fixups for the `Refund` timer ticks based on absolute expiry for consideration.

Also commit 1699aa646210ae7f0b1b1455ec63a5d3567ce514 for supporting variable number of timer ticks.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -691,6 +731,8 @@ fn packet_payloads_and_keys<T: CustomOnionMessageContents, S: secp256k1::Signing
 						next_blinding_override: Some(blinding_pt),
 					})), control_tlvs_ss));
+				} else {
+					prev_control_tlvs_ss = Some(control_tlvs_ss);
```
Yeah, looks like it is no longer needed after rebasing as the new code was added in 7c1726b585af71897e15dcb83024d5cab55d4fcd.

jkczyz commented:
> Speaking of timeouts, we should check if the payment has timed out per the `AwaitingInvoice::retry_stragety` (assuming its a time-based one) in the timer tick too.

Discussed offline. It would be better to keep the retry strategy separate from when to expire waiting for an invoice.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1216,6 +1223,8 @@ where
 	event_persist_notifier: Notifier,
 	needs_persist_flag: AtomicBool,
 
+	pending_offers_messages: Mutex<Vec<PendingOnionMessage<OffersMessage>>>,
```
Moved these to #2039.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7071,6 +7093,172 @@ where
+
+		let builder = RefundBuilder::deriving_payer_id(
+			description, node_id, expanded_key, entropy, secp_ctx, amount_msats, payment_id
+		)?
```
I'm going to leave this as-is. We have a similar problem in `pay_for_offer` only there are many more ways the builder can fail there.

valentinewallace reviewed (commented):
LGTM after rebase

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7071,6 +7090,71 @@ where
+
+		self.pending_outbound_payments
+			.add_new_awaiting_invoice(
+				payment_id, absolute_expiry, retry_strategy, max_total_routing_fee_msat,
```
This is already supported more generally by `ChannelManager::abandon_payment`.

jkczyz commented:
> Sorry needs a small-ish rebase after #2599.

Rebased. Needed to expose `ParsedOnionMessageContents` because of this.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7108,6 +7127,71 @@ where
+	/// [`ChannelManager`] when handling [`Bolt12Invoice`] messages for the refund. The builder will
+	/// have the provided expiration set. Any changes to the expiration on the returned builder will
+	/// not be honored by [`ChannelManager`].
+	///
```
Updated in #2039.
